README.md
# Cardinal Numerals

![Cardinal Numbers (Tungstenblue/Getty Images)](cardinal_numbers.png)

A [cardinal numeral](<https://en.wikipedia.org/wiki/Cardinal_number_(linguistics)>) is a noun used to count, such as the Vietnamese words **một**, **hai**, **ba**, but also compounds, such as **ba trăm bốn mươi hai**.

# Waypoint 1: Generate Vietnamese Numerals

Write a function `integer_to_vietnamese_numeral` that takes an argument `n` representing a positive integer, and returns a string corresponding to the **North** [Vietnamese cardinal numeral](https://en.wikipedia.org/wiki/Vietnamese_numerals) of this number.

The function raises the exception `TypeError` if the argument `n` is not an integer.

The function raises the exception `ValueError` if the argument `n` is not positive integer.

The maximum integer the function can generate a cardinal numeral is `999,999,999,999`. If `n` is over this number, the function raises the exception `OverflowError`.

For example:

```python
>>> integer_to_vietnamese_numeral(96)
'chín mươi sáu'
>>> integer_to_vietnamese_numeral(405)
'bốn trăm linh năm'
>>> integer_to_vietnamese_numeral(1915)
'một nghìn chín trăm mười lăm'
>>> integer_to_vietnamese_numeral(5061)
'năm nghìn không trăm sáu mươi mốt'
>>> integer_to_vietnamese_numeral(1002003)
'một triệu không trăm linh hai nghìn không trăm linh ba'
>>> integer_to_vietnamese_numeral(1000000)
'một triệu'
>>> integer_to_vietnamese_numeral(1030000)
'một triệu không trăm ba mươi nghìn'
>>> integer_to_vietnamese_numeral(1002003)
'một triệu không trăm linh hai nghìn không trăm linh ba'
>>> integer_to_vietnamese_numeral(1002003004)
'một tỷ không trăm linh hai triệu không trăm linh ba nghìn không trăm linh bốn'
>>> integer_to_vietnamese_numeral(1002003004)
'một tỷ không trăm linh ba nghìn không trăm linh bốn'
>>> integer_to_vietnamese_numeral(1002000004)
'một tỷ không trăm linh hai triệu không trăm linh bốn'
>>> integer_to_vietnamese_numeral(100000004)
'một trăm triệu không trăm linh bốn'
>>> integer_to_vietnamese_numeral(999999999999)
'chín trăm chín mươi chín tỷ chín trăm chín mươi chín triệu chín trăm chín mươi chín nghìn chín trăm chín mươi chín'
>>> integer_to_vietnamese_numberal('9999999999990')
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "<input>", line 4, in _integer_to_vietnamese_numeral
Overflow: Integer greater than 999,999,999,999
>>> integer_to_vietnamese_numberal('4096')
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "<input>", line 3, in _integer_to_vietnamese_numeral
TypeError: Not an integer
>>> integer_to_vietnamese_numeral(-1)
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "<input>", line 6, in _integer_to_vietnamese_numeral
ValueError: Not a positive integer
```

_Hint: if you have not a clear idea how to start, you should consider revising the [international system of numeration](https://www.youtube.com/watch?v=jKo_NrYcAaA), which is followed by most of the countries of the world, including Vietnam. In this system, a number is split up into groups (periods). There is a pattern that repeats for each group._

![International System of Numeration](international_system_of_numeration.png)

# Waypoint 2: Support Vietnamese Southern Numerals

The words "nghìn" and "linh" are the standard reading in Northern Vietnam, whilst "ngàn" and "lẻ" are more commonly used in the South.

Add an [optional argument](http://www.yourownlinux.com/2016/12/python-function-keyword-defaults-optional-arguments.html) `region` (a string) to the function `integer_to_vietnamese_numeral` that can be passed with one of the two possible values (strings):

- `'north'`: Northern Vietnam
- `'south'`: Southern Vietnam

By default, the parameter `region` is set to `north`. If the argument `region` is passed to the function `integer_to_vietnamese_numeral` with the value `None`, it defaults to `north`.

If the argument `region` passed to the function is not `None` and is not a string, the function `integer_to_vietnamese_numeral` raises an exception `TypeError`.

If the argument `region` passed to the function is not `None` and is not one of the two values `'north'` and `'south'`, the function `integer_to_vietnamese_numeral` raises an exception `ValueError`.

For examples:

```python
>>> integer_to_vietnamese_numeral(405)
'bốn trăm linh năm'
>>> integer_to_vietnamese_numeral(405, region='south')
'bốn trăm lẻ năm'
>>> integer_to_vietnamese_numeral(1971, region='north')
'một nghìn chín trăm bảy mươi một'
>>> integer_to_vietnamese_numeral(1971, region='south')
'một ngàn chín trăm bảy mươi một'
>>> integer_to_vietnamese_numeral(1, region=0)
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "<input>", line 3, in _integer_to_vietnamese_numeral
TypeError: Argument "region" is not a string
>>> integer_to_vietnamese_numeral(1, region='whatever')
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "<input>", line 6, in _integer_to_vietnamese_numeral
ValueError: Argument "region" has not a correct value
```

# Waypoint 3: Say Vietnamese Numeral

We would like our function `integer_to_vietnamese_numeral` to say the Vietnamese numeral that the function generates from the argument `n`. This is called [text-to-speech (TTS)](https://en.wikipedia.org/wiki/Speech_synthesis). Text-to-speech is [commonly used nowadays on mobile phones](https://www.youtube.com/watch?v=-gTn4Q-9Lk8).

Our implementation of the generation of synthesized speech from Vietnamese numerals will be simple. The function `integer_to_vietnamese_numeral` will successively play the sound of each Vietnamese numeral that the function has generated from the argument `n`.

The directory `./sounds/vie/north` contains several sound files corresponding to each Vietnamese numeral:

```bash
$ ls -la ./sounds/vie/north
total 152
drwxr-xr-x  21 dcaune  staff   672 Jun 24 10:25 .
drwxr-xr-x   5 dcaune  staff   160 May 27 08:52 ..
-rw-r--r--@  1 dcaune  staff  2496 Jun 24 08:52 ba.mp3
-rw-r--r--@  1 dcaune  staff  2784 Jun 24 08:54 bảy.mp3
-rw-r--r--@  1 dcaune  staff  2784 Jun 24 08:53 bốn.mp3
-rw-r--r--@  1 dcaune  staff  2880 Jun 24 08:55 chín.mp3
-rw-r--r--@  1 dcaune  staff  2880 Jun 24 08:52 hai.mp3
-rw-r--r--@  1 dcaune  staff  2688 Jun 24 10:24 không.mp3
-rw-r--r--@  1 dcaune  staff  2304 Jun 24 08:54 lăm.mp3
-rw-r--r--@  1 dcaune  staff  2112 Jun 24 08:52 mốt.mp3
-rw-r--r--@  1 dcaune  staff  2016 Jun 24 08:51 một.mp3
-rw-r--r--@  1 dcaune  staff  2976 Jun 24 08:56 mươi.mp3
-rw-r--r--@  1 dcaune  staff  2976 Jun 24 08:56 mười.mp3
-rw-r--r--@  1 dcaune  staff  2304 Jun 24 08:53 năm.mp3
-rw-r--r--@  1 dcaune  staff  2592 Jun 24 08:57 ngàn.mp3
-rw-r--r--@  1 dcaune  staff  2400 Jun 24 08:57 nghìn.mp3
-rw-r--r--@  1 dcaune  staff  3264 Jun 24 08:54 sáu.mp3
-rw-r--r--@  1 dcaune  staff  2592 Jun 24 08:55 tám.mp3
-rw-r--r--@  1 dcaune  staff  2496 Jun 24 08:57 trăm.mp3
-rw-r--r--@  1 dcaune  staff  2688 Jun 24 08:58 triệu.mp3
-rw-r--r--@  1 dcaune  staff  2208 Jun 24 08:58 tỷ.mp3
```

For example, the Vietnamese numeral for `128` is "một trăm hai mươi tám". If we pass the argument `n` with the value `128`, the function `integer_to_vietnamese_numeral` needs to play the following sound files in that order:

1. `một.mp3`
1. `trăm.mp3`
1. `hai.mp3`
1. `mươi.mp3`
1. `tám.mp3`

Update the function `integer_to_vietnamese_numeral` by adding an optional argument `activate_tts` (a boolean), with the default value `False`.

If the argument `activacte_tts` is passed with the value `None`, it defaults to `False`.

When this argument `activate_tts` is `True`, the function says the Vietnamese numeral that the function generates from the argument `n`, and it returns the Vietnamese numeral. If this argument `activate_tts` is `True`, the function only returns the Vietnamese numeral without saying it.

If the argument `activate_tts` passed to the function is not `None` and it is not a boolean value, the function `integer_to_vietnamese_numeral` raises an exception `TypeError`.

For example:

```python
>>> integer_to_vietnamese_numeral(405, activate_tts=True)
'bốn trăm linh năm'
>>> integer_to_vietnamese_numeral(405, region='south', activate_tts=True)
'bốn trăm lẻ năm'
>>> integer_to_vietnamese_numeral(1971, activate_tts=True, region='north')
'một nghìn chín trăm bảy mươi một'
>>> integer_to_vietnamese_numeral(1971, region='south', activate_tts=True)
'một ngàn chín trăm bảy mươi một'
>>> integer_to_vietnamese_numeral(1, activate_tts=1)
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "<input>", line 3, in _integer_to_vietnamese_numeral
TypeError: Argument "activate_tts" is not a boolean
```

# Waypoint 4: Say Vietnamese Numeral with Southern Accent

The sound files contained in the directory `./sounds/vie` correspond to Vietnamese numerals said with the Northern accent. We would like to support Vietnamese Southern accent.

If the argument `region` is `'north'`, then the function `integer_to_vietnamese_numeral` says the numeral with the North accent, but if the argument `region` is `'south'`, then the function says the numeral with the South accent.

Unfortunately, we don't have sound files for Vietnamese Southern pronunciation of Vietnamese numerals. You need to create these audio files by recording yourself. You probably have an Android or a iPhone mobile phone. You will easily find a free voice recorder application on your favorite online store.

# Waypoint 5: Generate English Numerals

Write a function `integer_to_english_numeral` that takes an argument `n` representing a positive integer, and returns a string corresponding to the [English cardinal numeral](https://en.wikipedia.org/wiki/English_numerals) of this number.

The function raises the exception `TypeError` if the argument `n` is not an integer.

The function raises the exception `ValueError` if the argument `n` is not positive integer.

For example:

```python
>>> integer_to_english_numeral(96)
'ninety-six'
>>> integer_to_english_numeral(101)
'one hundred and one'
>>> integer_to_english_numeral(405)
'four hundred five'
>>> integer_to_english_numeral(1971)
'one thousand and nine hundred and seventy-one'
>>> integer_to_english_numeral(5061)
'five thousand and sixty-one'
>>> integer_to_english_numberal('4096')
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "<input>", line 3, in _integer_to_english_numeral
TypeError: Not an integer
>>> integer_to_english_numeral(-1)
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "<input>", line 6, in _integer_to_english_numeral
ValueError: Not a positive integer
```

_Note: You can check whether your function `integer_to_english_numeral` returns the correct English cardinal numeral by comparing this result with [online generators](https://www.ego4u.com/en/cram-up/vocabulary/numbers/generator)._

# Waypoint 6: Say English Numerals

We would also want our function `integer_to_english_numeral` to say the English numeral that the function generates from the argument `n`! 😁

Update the function `integer_to_english_numeral` by adding an optional argument `activate_tts` (a boolean), with the default value `False`. If the argument`activate_tts`is passed with the value`None`, it defaults to`False`.

When this argument `activate_tts` is `True`, the function says the English numeral generated from the argument `n`, and the function returns the English numeral. If this argument `activate_tts` is `True`, the function only returns the English numeral without saying it.

If the argument `activate_tts` passed to the function is not `None` and it is not a boolean value, the function raises an exception `TypeError`.

Where are the audio files for the English cardinal numerals? Guess what, you will also have to record these audio files yourself and save them in the directory `./sounds/eng`... Like a pro! 😈😎


CHANGELOG.md
# CHANGELOG.md

## Version 0.1 2019-07-16

Features:

- Generate north and south Vietnamese cardinal numeral of number and text-to-speech it

- Generate English cardinal numeral of number and text-to-speech it


## Version 0.2 2019-07-17

Fix:

- Modified time-play of audio to pronounce quite better

- Replaced south audio set


## Version 1.0 2019-07-18 (Release)

Features:

- Generate north and south Vietnamese cardinal numeral of number and text-to-speech it

- Generate English cardinal numeral of number and text-to-speech it


## Version 1.1 2019-07-18 (Release)

Fix:

- Corrected name file 'forty.ogg'


## Version 1.2 2019-07-18 (Release)

Fix:

- Fixed code in function integer_to_english_numeral()


## Version 1.3 2019-07-19 (Release)

Fix:

- Fix input case n = 0

- Optimized code


src/Test.py
import cardinal_numeral

print(cardinal_numeral.integer_to_vietnamese_numeral(583729193902, region = 'south', activate_tts = True))


src/test.py
import cardinal_numeral as cn
print(cn.integer_to_vietnamese_numeral(906030201, region = 'south', activate_tts = True))
print(cn.integer_to_vietnamese_numeral(906030201, region = 'north', activate_tts = True))
print(cn.integer_to_english_numeral(906030201, activate_tts = True))


src/cardinal_numeral_BACKUP_10856.py
# Defined numbers
NUMERAL_ZERO    =   "không"
NUMERAL_ONE1    =   "một"
NUMERAL_ONE2    =   "mốt"
NUMERAL_TWO     =   "hai"
NUMERAL_THREE   =   "ba"
NUMERAL_FOUR    =   "bốn"
NUMERAL_FIVE1   =   "năm"
NUMERAL_FIVE2   =   "lăm"
NUMERAL_SIX     =   "sáu"
NUMERAL_SEVEN   =   "bảy"
NUMERAL_EIGHT   =   "tám"
NUMERAL_NINE    =   "chín"
NUMERAL_TEN1    =   "mười"
NUMERAL_TEN2    =   "mươi"
ONES1           =   "linh"
ONES2           =   "lẻ"
HUNDREDS        =   "trăm"
THOUSANDS1      =   "nghìn"
THOUSANDS2      =   "ngàn"
MILLIONS        =   "triệu"
BILLIONS        =   "tỷ"

# User defined custom exception
class Overflow(Exception):
    """Integer greater than 999,999,999,999"""
    pass


def convert_units(n):
    """
    Return list cardinal numeral words from number of units.

    Parameters:
    n (int): n is a number between 1 and 9.

    Returns:
    res (list): res has 1 element, pronounce a number between 1 and 9.
    """

    res = []
    switcher = {
            0: NUMERAL_ZERO,
            1: NUMERAL_ONE1,
            2: NUMERAL_TWO,
            3: NUMERAL_THREE,
            4: NUMERAL_FOUR,
            5: NUMERAL_FIVE1,
            6: NUMERAL_SIX,
            7: NUMERAL_SEVEN,
            8: NUMERAL_EIGHT,
            9: NUMERAL_NINE
    }
    res.append(switcher[n])
    return res


def convert_tens(n):
    res = []
    """
    Return list cardinal numeral words from number of tens.

    Parameters:
    n (int): n is a number between 10 and 99.

    Returns:
    res (list): res have many element, pronounce a number between 10 and 99.
    """

    if n // 10 == 1:
        res.append(NUMERAL_TEN1)
    # n from 2x to 9x
    else:
        switcher = {
                2: NUMERAL_TWO,
                3: NUMERAL_THREE,
                4: NUMERAL_FOUR,
                5: NUMERAL_FIVE1,
                6: NUMERAL_SIX,
                7: NUMERAL_SEVEN,
                8: NUMERAL_EIGHT,
                9: NUMERAL_NINE
        }
        res.append(switcher[n // 10])
        res.append(NUMERAL_TEN2)

    # Convert digit of units
    # n % 10 != 0, so digit of units is from 1 to 9
    if n % 10 != 0:
        # n % 10 != 1, so digit of units is from 2 to 9
        if n % 10 != 1:
            switcher = {
                    2: NUMERAL_TWO,
                    3: NUMERAL_THREE,
                    4: NUMERAL_FOUR,
                    5: NUMERAL_FIVE2,
                    6: NUMERAL_SIX,
                    7: NUMERAL_SEVEN,
                    8: NUMERAL_EIGHT,
                    9: NUMERAL_NINE
            }
            res.append(switcher[n % 10])
        # In this case, the digit is definitely 1, một or mốt
        else:
            if n // 10 == 1:
                res.append(NUMERAL_ONE1)
            else:
                res.append(NUMERAL_ONE2)
    return res


def convert_hundreds(n, region = 'north'):
    """
    Return list cardinal numeral words from number of hundreds.

    Parameters:
    n (int): n is a number between 100 and 999, (can get value less than 100 to pronounce "khong tram").
    region (str): region is 'north' or 'south'.

    Returns:
    res (list): res have many element, pronounce a number between 100 and 999 (or less than 100 to pronounce "khong tram").
    """

    # waypoint 2, linh in north, lẻ in south
    reg = {'north': ONES1, 'south': ONES2}
    res = []
    res = res + convert_units(n // 100)
    res.append(HUNDREDS)

    if n % 100 != 0:
        if n % 100 >= 10:
            res = res + convert_tens(n % 100)
        else:
            res.append(reg[region])
            res = res + convert_units(n % 100)
    return res


# Waypoint 1
def integer_to_vietnamese_numeral(n, region = 'north', activate_tts = False):
    """
    Returns a string corresponding to Vietnamese cardinal numeral of a number.

    Parameters:
    n (int): n is a number maximum 999,999,999,999.
    region (str): region is 'north' or 'south'.
    activate_tts (bool): True with sound, False without sound.

    Returns:
    res (str): a string corresponding to Vietnamese cardinal numeral of a number.
    """

    reg = {'north': THOUSANDS1, 'south': THOUSANDS2}
    if (type(activate_tts) is not bool) and (type(activate_tts) is not None):
        raise TypeError('Argument "activate_tts" is not a boolean')
    if activate_tts is None:
        activate_tts = False
    if type(region) != str:
        raise TypeError('Argument "region" is not a string')
    if not region in reg:
        raise ValueError('Argument "region" has not a correct value')

    res = []
    # Catch exceptions
    if type(n) is not int:
        raise TypeError("Not an integer")
    if n < 0:
        raise ValueError("Not a positive integer")
    if n > 999999999999:
        raise Overflow("Integer greater than 999,999,999,999")

    list_divisor = (1000000000, 1000000, 1000, 1)
    for i in list_divisor:
        if n // i > 0:
            if len(res) == 0:
                if n // i >= 100:
                    res = res + convert_hundreds(n // i, region)
                elif n // i >= 10:
                    res = res + convert_tens(n // i)
                else:
                    res = res + convert_units(n // i)
            else:
                res = res + convert_hundreds(n // i, region)

            if i == 1000000000:
                res.append(BILLIONS)
            elif i == 1000000:
                res.append(MILLIONS)
            elif i == 1000:
                res.append(reg[region])

            n = n % i

    if not activate_tts:
        return ' '.join(res)
    else:
        import pygame
        pygame.init()
        for word in res:
            sound = pygame.mixer.Sound("../sounds/vie/" + str(region) + "/" + word + ".ogg")
            sound.play(maxtime = 3000)
            print("../sounds/vie/" + str(region) + "/" + word + ".ogg")
            pygame.time.delay(500)
        return ' '.join(res)

##################################################################

def convert_from_0_to_99(n):
    """
    Returns a string corresponding to a number between 0 and 99.
 
    Parameters:
    n (int): n is a number between 0 and 99.

    Returns:
    res (str): a string corresponding to English cardinal numeral of a number.
    """

    if n < 20:
        switcher = {
                0: "zero",
                1: "one",
                2: "two",
                3: "three",
                4: "four",
                5: "five",
                6: "six",
                7: "seven",
                8: "eight",
                9: "nine",
                10: "ten",
                11: "eleven",
                12: "twelve",
                13: "thirteen",
                14: "fourteen",
                15: "fifteen",
                16: "sixteen",
                17: "seventeen",
                18: "eighteen",
                19: "nineteen"
        }
        return switcher[n]
    else:
        switcher = {
                2: "twenty",
                3: "thirty",
                4: "forty",
                5: "fifty",
                6: "sixty",
                7: "seventy",
                8: "eighty",
                9: "ninety"
        }

        if n % 10 == 0:
            res = switcher[n // 10]
        else:
            res = switcher[n // 10] + "-" + convert_from_0_to_99(n % 10)

        return res


def convert_hundreds_en(n):
    """
    Returns a string corresponding to a number between 100 and 999.
 
    Parameters:
    n (int): n is a number between 100 and 999.

    Returns:
    res (list): a list of string corresponding to English cardinal numeral of a number.
    """

    res = []

    if n < 100:
        res.append(convert_from_0_to_99(n))
    else:
        res.append(convert_from_0_to_99(n // 100))
        res.append("hundred")

        if n % 100 != 0:
            res.append("and")
            res.append(convert_from_0_to_99(n % 100))

    return res


<<<<<<< HEAD
def integer_to_english_numeral(n, activate_tts = False):
=======
def integer_to_english_numeral(n):
    """
    Returns a string corresponding to the English cardinal numeral of this number. 

    Parameters:
    n (int): n is a number.
    activate_tts (bool): True with sound, False without sound.

    Returns:
    res (str): a string corresponding to English cardinal numeral of a number.
    """

>>>>>>> document_code
    if type(n) is not int:
        raise TypeError("Not an integer")
    if n < 0:
        raise ValueError("Not a positive integer")
    if n > 999999999999:
        raise Overflow("Integer greater than 999,999,999,999")

    if (type(activate_tts) is not bool) and (type(activate_tts) is not None):
        raise TypeError('Argument "activate_tts" is not a boolean')
    if activate_tts is None:
        activate_tts = False

    res = []
    list_divisor = (1000000000, 1000000, 1000, 1)
    for i in list_divisor:
        if n // i > 0:
            res = res + convert_hundreds_en(n // i)

            if i == 1000000000:
                res.append("billion")
            elif i == 1000000:
                res.append("million")
            elif i == 1000:
                res.append("thousand")

            if n % i != 0:
                res.append("and")

            n = n % i

    if not activate_tts:
        return ' '.join(res)
    else:
        # import pygame
        # pygame.init()
        for word in res:
            if word.find('-') == -1:
                # sound = pygame.mixer.Sound("../sounds/en/" + word + ".ogg")
                # sound.play(maxtime = 3000)
                print("../sounds/en/" + word + ".ogg")
                # pygame.time.delay(500)
            else:
                # sound = pygame.mixer.Sound("../sounds/en/" + word[: word.find('-')] + ".ogg")
                # sound.play(maxtime = 3000)
                print("../sounds/en/" + word[0 : word.find('-')] + ".ogg")
                # pygame.time.delay(500)

                # sound = pygame.mixer.Sound("../sounds/en/" + word[word.find('-') + 1 :] + ".ogg")
                # sound.play(maxtime = 3000)
                print("../sounds/en/" + word[word.find('-') + 1 :] + ".ogg")
                # pygame.time.delay(500)

        return ' '.join(res)


src/cardinal_numeral.py
# Defined path to audio's directory
SOUNDS_PATH    =   "../sounds/"

# Defined numbers
NUMERAL_ZERO    =   "không"
NUMERAL_ONE1    =   "một"
NUMERAL_ONE2    =   "mốt"
NUMERAL_TWO     =   "hai"
NUMERAL_THREE   =   "ba"
NUMERAL_FOUR    =   "bốn"
NUMERAL_FIVE1   =   "năm"
NUMERAL_FIVE2   =   "lăm"
NUMERAL_SIX     =   "sáu"
NUMERAL_SEVEN   =   "bảy"
NUMERAL_EIGHT   =   "tám"
NUMERAL_NINE    =   "chín"
NUMERAL_TEN1    =   "mười"
NUMERAL_TEN2    =   "mươi"
ONES1           =   "linh"
ONES2           =   "lẻ"
HUNDREDS        =   "trăm"
THOUSANDS1      =   "nghìn"
THOUSANDS2      =   "ngàn"
MILLIONS        =   "triệu"
BILLIONS        =   "tỷ"

# User defined custom exception
class Overflow(Exception):
    """
    Integer greater than 999,999,999,999
    """
    pass


def pronounce_word(path):
    """
    Function to play audio file from path directory
    """
    import pygame
    pygame.init()
    sound = pygame.mixer.Sound(path)
    sound.play()
    pygame.time.delay(600)


def convert_units(n):
    """
    Return list cardinal numeral words from number of units.

    Parameters:
        n (int): n is a number from 0 to 9.

    Returns:
        cardinal_numeral_list (list): list of cardinal numeral words.
    """

    cardinal_numeral_list = []
    switcher = {
            0: NUMERAL_ZERO,
            1: NUMERAL_ONE1,
            2: NUMERAL_TWO,
            3: NUMERAL_THREE,
            4: NUMERAL_FOUR,
            5: NUMERAL_FIVE1,
            6: NUMERAL_SIX,
            7: NUMERAL_SEVEN,
            8: NUMERAL_EIGHT,
            9: NUMERAL_NINE
    }
    cardinal_numeral_list.append(switcher[n])

    return cardinal_numeral_list


def convert_tens(n):
    """
    Return list cardinal numeral words from number of tens.

    Parameters:
        n (int): n is a number from 10 to 99.

    Returns:
        cardinal_numeral_list (list): list of cardinal numeral words
    """

    cardinal_numeral_list = []

    # Converting digit of tens
    # In case digit is 1
    if n // 10 == 1:
        cardinal_numeral_list.append(NUMERAL_TEN1)
    # In case digit from 2 to 9
    else:
        switcher = {
                2: NUMERAL_TWO,
                3: NUMERAL_THREE,
                4: NUMERAL_FOUR,
                5: NUMERAL_FIVE1,
                6: NUMERAL_SIX,
                7: NUMERAL_SEVEN,
                8: NUMERAL_EIGHT,
                9: NUMERAL_NINE
        }
        cardinal_numeral_list.append(switcher[n // 10])
        cardinal_numeral_list.append(NUMERAL_TEN2)

    # Convert digit of units
    if n % 10 != 0:
        # Digit is from 2 to 9
        if n % 10 != 1:
            switcher = {
                    2: NUMERAL_TWO,
                    3: NUMERAL_THREE,
                    4: NUMERAL_FOUR,
                    5: NUMERAL_FIVE2,
                    6: NUMERAL_SIX,
                    7: NUMERAL_SEVEN,
                    8: NUMERAL_EIGHT,
                    9: NUMERAL_NINE
            }
            cardinal_numeral_list.append(switcher[n % 10])
        # Digit is 1
        else:
            if n // 10 == 1:
                cardinal_numeral_list.append(NUMERAL_ONE1)
            else:
                cardinal_numeral_list.append(NUMERAL_ONE2)

    return cardinal_numeral_list


def convert_hundreds(n, region = 'north'):
    """
    Return list cardinal numeral words from number of hundreds.

    Parameters:
        n (int): n is a number from 100 to 999, (can get value less than 100 to pronounce "khong tram").
        region (str): region is 'north' or 'south'.

    Returns:
        cardinal_numeral_list (list): list of cardinal numeral words.
    """

    reg = {'north': ONES1, 'south': ONES2}
    cardinal_numeral_list = []
    cardinal_numeral_list += convert_units(n // 100)
    cardinal_numeral_list.append(HUNDREDS)

    if n % 100 != 0:
        if n % 100 >= 10:
            cardinal_numeral_list += convert_tens(n % 100)
        else:
            cardinal_numeral_list.append(reg[region])
            cardinal_numeral_list += convert_units(n % 100)

    return cardinal_numeral_list


def integer_to_vietnamese_numeral(n, region = 'north', activate_tts = False):
    """
    Returns a string corresponding to Vietnamese cardinal numeral of a number.

    Parameters:
        n (int): n is a number from 0 to 999,999,999,999.
        region (str): region gets 'north' or 'south' to generate north or south Vietnamese cardinal numeral.
        activate_tts (bool): activate_tts gets True or False to say it or not.

    Returns:
        (str): a string corresponding to Vietnamese cardinal numeral of a number.
    """

    if (type(activate_tts) is not bool) and (type(activate_tts) is not None):
        raise TypeError('Argument "activate_tts" is not a boolean')
    if activate_tts is None:
        activate_tts = False

    reg = {'north': THOUSANDS1, 'south': THOUSANDS2}
    if type(region) is not str:
        raise TypeError('Argument "region" is not a string')
    if not region in reg:
        raise ValueError('Argument "region" has not a correct value')

    if type(n) is not int:
        raise TypeError("Not an integer")
    if n < 0:
        raise ValueError("Not a positive integer")
    if n > 999999999999:
        raise Overflow("Integer greater than 999,999,999,999")

    cardinal_numeral_list = []
    # Devide n to many set of three numbers
    list_divisor = (1000000000, 1000000, 1000, 1)
    for i in list_divisor:
        # If set of three numbers has any number is not zero
        if n // i > 0:
            # Saying if it is the first set of three numbers
            if len(cardinal_numeral_list) == 0:
                if n // i >= 100:
                    cardinal_numeral_list += convert_hundreds(n // i, region)
                elif n // i >= 10:
                    cardinal_numeral_list += convert_tens(n // i)
                else:
                    cardinal_numeral_list += convert_units(n // i)
            # Saying if it is NOT the first set of three numbers
            else:
                cardinal_numeral_list += convert_hundreds(n // i, region)

            # Adding place-value for set of three numbers
            if i == 1000000000:
                cardinal_numeral_list.append(BILLIONS)
            elif i == 1000000:
                cardinal_numeral_list.append(MILLIONS)
            elif i == 1000:
                cardinal_numeral_list.append(reg[region])

            # Removing set of three numbers converted
            n = n % i

    # If cardinal_numeral_list still empty, this is in case n = 0
    if len(cardinal_numeral_list) == 0:
        cardinal_numeral_list += convert_units(0)

    # Check activate_tts
    if activate_tts:
        for word in cardinal_numeral_list:
            pronounce_word(SOUNDS_PATH + 'vie/' + region + "/" + word + ".ogg")
            print(SOUNDS_PATH + 'vie/' + region + "/" + word + ".ogg")

    return ' '.join(cardinal_numeral_list)


################################################################################################################


def convert_from_0_to_99(n):
    """
    Returns a string corresponding to a number between 0 and 99.
 
    Parameters:
        n (int): n is a number from 0 to 99.

    Returns:
        (str): a string corresponding to English cardinal numeral of a number.
    """

    if n < 20:
        switcher = {
                0: "zero",
                1: "one",
                2: "two",
                3: "three",
                4: "four",
                5: "five",
                6: "six",
                7: "seven",
                8: "eight",
                9: "nine",
                10: "ten",
                11: "eleven",
                12: "twelve",
                13: "thirteen",
                14: "fourteen",
                15: "fifteen",
                16: "sixteen",
                17: "seventeen",
                18: "eighteen",
                19: "nineteen"
        }
        return switcher[n]
    else:
        switcher = {
                2: "twenty",
                3: "thirty",
                4: "forty",
                5: "fifty",
                6: "sixty",
                7: "seventy",
                8: "eighty",
                9: "ninety"
        }

        if n % 10 == 0:
            return switcher[n // 10]

        return switcher[n // 10] + "-" + convert_from_0_to_99(n % 10)


def convert_hundreds_en(n):
    """
    Returns a string corresponding to a number from 100 to 999
 
    Parameters:
        n (int): n is a number from 100 to 999.

    Returns:
        cardinal_numeral_list (list): a list of words corresponding to English cardinal numeral of a number.
    """

    cardinal_numeral_list = []

    if n < 100:
        cardinal_numeral_list.append(convert_from_0_to_99(n))
    else:
        cardinal_numeral_list.append(convert_from_0_to_99(n // 100))
        cardinal_numeral_list.append("hundred")

        if n % 100 != 0:
            cardinal_numeral_list.append("and")
            cardinal_numeral_list.append(convert_from_0_to_99(n % 100))

    return cardinal_numeral_list


def integer_to_english_numeral(n, activate_tts = False):
    """
    Returns a string corresponding to the English cardinal numeral of this number. 

    Parameters:
        n (int): n is a number.
        activate_tts (bool): True with sound, False without sound.

    Returns:
        cardinal_numeral_list (list): a list of words corresponding to English cardinal numeral of a number.
    """

    if type(n) is not int:
        raise TypeError("Not an integer")
    if n < 0:
        raise ValueError("Not a positive integer")
    if n > 999999999999:
        raise Overflow("Integer greater than 999,999,999,999")

    if (type(activate_tts) is not bool) and (type(activate_tts) is not None):
        raise TypeError('Argument "activate_tts" is not a boolean')
    if activate_tts is None:
        activate_tts = False

    cardinal_numeral_list = []
    # Devide n to many set of three numbers
    list_divisor = (1000000000, 1000000, 1000, 1)
    for i in list_divisor:
        # If set of three numbers has any number is not zero
        if n // i > 0:
            cardinal_numeral_list += convert_hundreds_en(n // i)

            # Adding place-value for set of three numbers
            if i == 1000000000:
                cardinal_numeral_list.append("billion")
            elif i == 1000000:
                cardinal_numeral_list.append("million")
            elif i == 1000:
                cardinal_numeral_list.append("thousand")

            if n % i != 0:
                cardinal_numeral_list.append("and")

            # Removing set of three numbers converted
            n = n % i

    # If cardinal_numeral_list still empty, this is in case n = 0
    if len(cardinal_numeral_list) == 0:
        cardinal_numeral_list.append(convert_from_0_to_99(0))

    if activate_tts:
        for word in cardinal_numeral_list:
            if word.find('-') == -1:
                pronounce_word(SOUNDS_PATH + "en/" + word + ".ogg")
                print(SOUNDS_PATH + "en/" + word + ".ogg")
            else:
                # Play sound before '-'
                pronounce_word(SOUNDS_PATH + "en/" + word[: word.find('-')] + ".ogg")
                print(SOUNDS_PATH + "en/" + word[: word.find('-')] + ".ogg")

                # Play sound after '-'
                pronounce_word(SOUNDS_PATH + "en/" + word[word.find('-') + 1 :] + ".ogg")
                print(SOUNDS_PATH + "en/" + word[word.find('-') + 1 :] + ".ogg")

    return ' '.join(cardinal_numeral_list)
